### PR TITLE
fix: ensure batch values are sharable

### PR DESCRIPTION
**Pull request by wyattjoh**
Previously, the incremental cache returned objects with a readable stream (that were consumed). These can't be shared. Instead, this modifies the cache to instead return strings which are sharable.

This also fixes a bug related to revalidation when writing an updated prefetch RSC payload to the filesystem when PPR is enabled. Previously it just wrote to `.rsc` and now it correctly writes a `.prefetch.rsc` file instead.

### CODE REVIEW

**Comment by wyattjoh** [comment]
Current dependencies on/for this PR:
* canary
  * **PR #58315** <a href="https://app.graphite.dev/github/pr/vercel/next.js/58315?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/vercel/next.js/58315?utm_source=stack-comment).

File: packages/next/src/server/lib/incremental-cache/file-system-cache.ts [hunk 1]
```diff
@@ -339,7 +339,13 @@ export default class FileSystemCache implements CacheHandler {
...
       await this.fs.writeFile(
         this.getFilePath(
-          `${key}.${isAppPath ? 'rsc' : 'json'}`,
+          `${key}.${
+            isAppPath
+              ? this.experimental.ppr
+                ? RSC_PREFETCH_SUFFIX
+                : RSC_SUFFIX
+              : 'json'
+          }`,
```
Comment (wyattjoh) [inline packages/next/src/server/lib/incremental-cache/file-system-cache.ts:348]: This handles when we update the static RSC prefetch payload. Previously it just wrote a `.rsc` file when during PPR we instead have to write a `.prefetch.rsc` payload.

File: packages/next/src/server/response-cache/index.ts [hunk 2]
```diff
@@ -8,16 +8,16 @@ import type {
...
 import { scheduleOnNextTick } from '../../lib/scheduler'
+import { fromResponseCacheEntry, toResponseCacheEntry } from './utils'
 
 export * from './types'
 
 export default class ResponseCache implements ResponseCacheBase {
   private readonly batcher = Batcher.create<
     { key: string; isOnDemandRevalidate: boolean },
-    ResponseCacheEntry | null,
+    IncrementalCacheItem | null,
```
Comment (wyattjoh) [inline packages/next/src/server/response-cache/index.ts:20]: The `IncrementalCacheItem` is sharable as it uses a `string` for the `html` part and not a `RenderResult` which under the hood uses a `ReadableStream` which is consumed on read.

**Comment by ijjk** [comment]
## Failing test suites 
Commit: bd2a357b80be76e187950299e310ff4641c2c569

`pnpm test test/integration/image-optimizer/test/minimum-cache-ttl.test.ts` 
- with minimumCacheTTL of 5 sec > dev support with next.config.js > should use cache and stale-while-revalidate when query is the same for external image
- with minimumCacheTTL of 5 sec > dev support with next.config.js > should use cache and stale-while-revalidate when query is the same for internal image
- with minimumCacheTTL of 5 sec > dev support with next.config.js > should use cached image file when parameters are the same for animated gif
- with minimumCacheTTL of 5 sec > dev support with next.config.js > should handle concurrent requests

<details>
<summary>Expand output</summary>

  ● with minimumCacheTTL of 5 sec › dev support with next.config.js › should use cache and stale-while-revalidate when query is the same for external image

    expect(received).toBe(expected) // Object.is equality

    Expected: "MISS"
    Received: "HIT"

      749 |       expect(one.duration).toBeGreaterThan(delay)
      750 |       expect(one.res.status).toBe(200)
    > 751 |       expect(one.res.headers.get('X-Nextjs-Cache')).toBe('MISS')
          |                                                     ^
      752 |       expect(one.res.headers.get('Content-Type')).toBe('image/webp')
      753 |       expect(one.res.headers.get('Content-Disposition')).toBe(
      754 |         `${contentDispositionType}; filename="slow.webp"`

      at Object.toBe (integration/image-optimizer/test/util.ts:751:53)

  ● with minimumCacheTTL of 5 sec › dev support with next.config.js › should use cache and stale-while-revalidate when query is the same for internal image

    expect(received).toBe(expected) // Object.is equality

    Expected: "MISS"
    Received: "HIT"

      895 |     )
      896 |     expect(one.res.status).toBe(200)
    > 897 |     expect(one.res.headers.get('X-Nextjs-Cache')).toBe('MISS')
          |                                                   ^
      898 |     expect(one.res.headers.get('Content-Type')).toBe('image/webp')
      899 |     expect(one.res.headers.get('Content-Disposition')).toBe(
      900 |       `${contentDispositionType}; filename="test.webp"`

      at Object.toBe (integration/image-optimizer/test/util.ts:897:51)

  ● with minimumCacheTTL of 5 sec › dev support with next.config.js › should use cached image file when parameters are the same for animated gif

    expect(received).toBe(expected) // Object.is equality

    Expected: "MISS"
    Received: "HIT"

      1032 |     const res1 = await fetchViaHTTP(ctx.appPort, '/_next/image', query, opts)
      1033 |     expect(res1.status).toBe(200)
    > 1034 |     expect(res1.headers.get('X-Nextjs-Cache')).toBe('MISS')
           |                                                ^
      1035 |     expect(res1.headers.get('Content-Type')).toBe('image/gif')
      1036 |     expect(res1.headers.get('Content-Disposition')).toBe(
      1037 |       `${contentDispositionType}; filename="animated.gif"`

      at Object.toBe (integration/image-optimizer/test/util.ts:1034:48)

  ● with minimumCacheTTL of 5 sec › dev support with next.config.js › should handle concurrent requests

    expect(received).toEqual(expected) // deep equality

    - Expected  - 3
    + Received  + 3

      Array [
    -   "MISS",
    -   "MISS",
    -   "MISS",
    +   "HIT",
    +   "HIT",
    +   "HIT",
      ]

      1280 |       // until the cache be populated so all concurrent
      1281 |       // requests receive the same response
    > 1282 |       expect(xCache).toEqual(['MISS', 'MISS', 'MISS'])
           |                      ^
      1283 |     })
      1284 |   }
      1285 |

      at Object.toEqual (integration/image-optimizer/test/util.ts:1282:22)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js fix/incremental-cache-ppr | Change |
| - | - | - | - |
| buildDuration | 10.7s | 10.7s | N/A |
| buildDurationCached | 6.2s | 6.7s | ⚠️ +558ms |
| nodeModulesSize | 199 MB | 199 MB | ⚠️ +16 kB |
| nextStartRea..uration (ms) | 397ms | 399ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/incremental-cache-ppr | Change |
| - | - | - | - |
| 199-HASH.js gzip | 29.2 kB | 29.3 kB | N/A |
| 3f784ff6-HASH.js gzip | 53.3 kB | 53.3 kB | N/A |
| 494.HASH.js gzip | 180 B | 181 B | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 241 B | 239 B | N/A |
| main-HASH.js gzip | 31.7 kB | 31.8 kB | N/A |
| webpack-HASH.js gzip | 1.7 kB | 1.7 kB | ✓ |
| Overall change | 46.9 kB | 46.9 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/incremental-cache-ppr | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/incremental-cache-ppr | Change |
| - | - | - | - |
| _app-HASH.js gzip | 194 B | 195 B | N/A |
| _error-HASH.js gzip | 182 B | 181 B | N/A |
| amp-HASH.js gzip | 504 B | 506 B | N/A |
| css-HASH.js gzip | 322 B | 323 B | N/A |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | ✓ |
| edge-ssr-HASH.js gzip | 253 B | 255 B | N/A |
| head-HASH.js gzip | 348 B | 347 B | N/A |
| hooks-HASH.js gzip | 369 B | 368 B | N/A |
| image-HASH.js gzip | 4.3 kB | 4.3 kB | N/A |
| index-HASH.js gzip | 256 B | 256 B | ✓ |
| link-HASH.js gzip | 2.65 kB | 2.65 kB | N/A |
| routerDirect..HASH.js gzip | 311 B | 311 B | ✓ |
| script-HASH.js gzip | 384 B | 383 B | N/A |
| withRouter-HASH.js gzip | 307 B | 308 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 3.17 kB | 3.17 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/incremental-cache-ppr | Change |
| - | - | - | - |
| _buildManifest.js gzip | 483 B | 484 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/incremental-cache-ppr | Change |
| - | - | - | - |
| index.html gzip | 529 B | 527 B | N/A |
| link.html gzip | 541 B | 539 B | N/A |
| withRouter.html gzip | 524 B | 521 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/incremental-cache-ppr | Change |
| - | - | - | - |
| edge-ssr.js gzip | 92.5 kB | 92.5 kB | N/A |
| page.js gzip | 145 kB | 145 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/incremental-cache-ppr | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 626 B | 627 B | N/A |
| middleware-r..fest.js gzip | 150 B | 151 B | N/A |
| middleware.js gzip | 24.8 kB | 24.8 kB | N/A |
| edge-runtime..pack.js gzip | 1.92 kB | 1.92 kB | ✓ |
| Overall change | 1.92 kB | 1.92 kB | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js fix/incremental-cache-ppr | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 167 kB | 167 kB | ✓ |
| app-page-exp..prod.js gzip | 93.2 kB | 93.2 kB | ✓ |
| app-page-tur..prod.js gzip | 93.9 kB | 93.9 kB | ✓ |
| app-page-tur..prod.js gzip | 88.5 kB | 88.5 kB | ✓ |
| app-page.run...dev.js gzip | 137 kB | 137 kB | ✓ |
| app-page.run..prod.js gzip | 87.9 kB | 87.9 kB | ✓ |
| app-route-ex...dev.js gzip | 23.8 kB | 23.8 kB | ✓ |
| app-route-ex..prod.js gzip | 16.4 kB | 16.4 kB | ✓ |
| app-route-tu..prod.js gzip | 16.4 kB | 16.4 kB | ✓ |
| app-route-tu..prod.js gzip | 16 kB | 16 kB | ✓ |
| app-route.ru...dev.js gzip | 23.2 kB | 23.2 kB | ✓ |
| app-route.ru..prod.js gzip | 16 kB | 16 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.37 kB | 9.37 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.64 kB | 9.64 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.37 kB | 9.37 kB | ✓ |
| pages-turbo...prod.js gzip | 21.8 kB | 21.8 kB | ✓ |
| pages.runtim...dev.js gzip | 22.5 kB | 22.5 kB | ✓ |
| pages.runtim..prod.js gzip | 21.8 kB | 21.8 kB | ✓ |
| server.runti..prod.js gzip | 48.8 kB | 49 kB | ⚠️ +163 B |
| Overall change | 922 kB | 923 kB | ⚠️ +163 B |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>page.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>edge-ssr.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>server.runtime.prod.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: bd2a357b80be76e187950299e310ff4641c2c569

**Review by ijjk — APPROVED** [review]

**Pull request merged**